internal/log/zap/zap.go

```diff
@@ -0,0 +1,222 @@
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package zap
+
+import (
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/log/format"
+	"github.com/vdaas/vald/internal/log/level"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+const (
+	detailsKey = "details"
+
+	defaultLevel = zapcore.DebugLevel
+)
+
+var (
+	zapcore_NewConsoleEncoder = zapcore.NewConsoleEncoder
+	zapcore_NewJSONEncoder    = zapcore.NewJSONEncoder
+)
+
+type logger struct {
+	format format.Format
+	level  level.Level
+
+	enableCaller bool
+
+	logger *zap.Logger
+	sugar  *zap.SugaredLogger
+}
+
+// New returns a new logger instance.
+func New(opts ...Option) (*logger, error) {
+	l := new(logger)
+	for _, opt := range append(defaultOpts, opts...) {
+		opt(l)
+	}
+
+	err := l.initialize("stdout", "stderr")
+	if err != nil {
+		return nil, err
+	}
+
+	return l, nil
+}
+
+func (l *logger) initialize(sinkPath, errSinkPath string) (err error) {
+	sink, closeOut, err := zap.Open(sinkPath)
+	if err != nil {
+		return err
+	}
+
+	errSink, _, err := zap.Open(errSinkPath)
+	if err != nil {
+		closeOut()
+		return err
+	}
+
+	core := zapcore.NewCore(
+		toZapEncoder(l.format),
+		sink,
+		toZapLevel(l.level),
+	)
+
+	opts := []zap.Option{
+		zap.ErrorOutput(errSink),
+	}
+
+	if l.enableCaller {
+		opts = append(opts, zap.AddCaller())
+	}
+
+	l.logger = zap.New(core, opts...)
+
+	l.sugar = l.logger.Sugar()
+
+	return nil
+}
+
+func (l *logger) Close() error {
+	err := l.logger.Sync()
+	if err != nil {
+		return errors.Wrap(l.sugar.Sync(), err.Error())
+	}
+
+	return l.sugar.Sync()
+}
+
+func toZapLevel(lv level.Level) zapcore.Level {
+	switch lv {
+	case level.DEBUG:
+		return zapcore.DebugLevel
+	case level.INFO:
+		return zapcore.InfoLevel
+	case level.WARN:
+		return zapcore.WarnLevel
+	case level.ERROR:
+		return zapcore.ErrorLevel
+	case level.FATAL:
+		return zapcore.FatalLevel
+	case level.Unknown:
+		fallthrough
+	default:
+		return defaultLevel
+	}
+}
+
+func toZapEncoder(fmt format.Format) zapcore.Encoder {
+	cfg := zap.NewProductionEncoderConfig()
+
+	switch fmt {
+	case format.RAW:
+		return zapcore_NewConsoleEncoder(cfg)
+	case format.JSON:
+		return zapcore_NewJSONEncoder(cfg)
+	case format.Unknown:
+		fallthrough
+	default:
+		return zapcore_NewJSONEncoder(cfg)
+	}
+}
+
+func (l *logger) Debug(vals ...interface{}) {
+	l.sugar.Debug(vals...)
+}
+
+func (l *logger) Debugf(format string, vals ...interface{}) {
+	l.sugar.Debugf(format, vals...)
+}
+
+func (l *logger) Debugd(msg string, details ...interface{}) {
+	if len(details) == 1 {
+		l.logger.Debug(msg, zap.Any(detailsKey, details[0]))
```

Comment:
do we need to unwrap the slice type if member equals 1.

Reply:
We should do it here because the given `details` is treated as slice and then zap will log it as the format like:
```
{"level": "debug", ... , "details": [{"foo": "bar", ...}]}
```
we want here is:
```
{"level": "debug", ... , "details": {"foo": "bar", ...}}
```

Reply:
Or should we revise the interface of `debugd` like below?
```
Debugd(msg string, details interface{})
```

Reply:
hmm, I think it's okay golang can't use override.
I understand about background, it's okay.